PR: https://github.com/odoo/odoo/pull/

From: e68ec01b1b33e8743d6103aecc92e91233c43bc4
From: Quentin Wolfs
Date: 2021-08-12 06:34:04

Structural Changes: 3
Total Changes: 40

[IMP] repair: add link from lot to repair orders

Add links to repair orders in which the tracked lot was consumed

Task-2467686

================================= pseudo patch: =================================

--- a/addons/repair/models/__init__.py
+++ b/addons/repair/models/__init__.py
@@ -3,4 +3,5 @@
 
 from . import repair
 from . import stock_traceability
+from . import stock_production_lot
 from . import account_move

--- a/None
+++ b/addons/repair/models/stock_production_lot.py
@@ -0,0 +1,39 @@
+# -*- coding: utf-8 -*-
+
+from collections import defaultdict
+from odoo import api, fields, models, _
+
+class ProductionLot(models.Model):
+    _inherit = 'stock.production.lot'
+
+    repair_order_ids = fields.Many2many('repair.order', string="Repair Orders", compute="_compute_repair_order_ids")
+    repair_order_count = fields.Integer('Repair order count', compute="_compute_repair_order_ids")
+
+    @api.depends('name')
+    def _compute_repair_order_ids(self):
+        repair_orders = defaultdict(lambda: self.env['repair.order'])
+        for repair_line in self.env['repair.line'].search([('lot_id', 'in', self.ids), ('state', '=', 'done')]):
+            repair_orders[repair_line.lot_id.id] |= repair_line.repair_id
+        for lot in self:
+            lot.repair_order_ids = repair_orders[lot.id]
+            lot.repair_order_count = len(lot.repair_order_ids)
+
+    def action_view_ro(self):
+        self.ensure_one()
+
+        action = {
+            'res_model': 'repair.order',
+            'type': 'ir.actions.act_window'
+        }
+        if len(self.repair_order_ids) == 1:
+            action.update({
+                'view_mode': 'form',
+                'res_id': self.repair_order_ids[0].id
+            })
+        else:
+            action.update({
+                'name': _("Repair orders of %s", self.name),
+                'domain': [('id', 'in', self.repair_order_ids.ids)],
+                'view_mode': 'tree,form'
+            })
+        return action
